skbio/diversity/tests/test_driver.py

```diff
@@ -0,0 +1,590 @@
+
+        # tree with no branch lengths
+        t = TreeNode.read(
+            StringIO(u'((((OTU1,OTU2),OTU3)),(OTU4,OTU5));'))
+        counts = [1, 2, 3]
+        otu_ids = ['OTU1', 'OTU2', 'OTU3']
+        self.assertRaises(ValueError, beta_diversity,
+                          'weighted_unifrac', counts, otu_ids=otu_ids, tree=t)
+        self.assertRaises(ValueError, beta_diversity,
+                          'unweighted_unifrac', counts, otu_ids=otu_ids,
+                          tree=t)
+
+        # tree missing some branch lengths
+        t = TreeNode.read(
+            StringIO(u'(((((OTU1,OTU2:0.5):0.5,OTU3:1.0):1.0):0.0,(OTU4:'
+                     u'0.75,OTU5:0.75):1.25):0.0)root;'))
+        counts = [1, 2, 3]
+        otu_ids = ['OTU1', 'OTU2', 'OTU3']
+        self.assertRaises(ValueError, beta_diversity,
+                          'weighted_unifrac', counts, otu_ids=otu_ids, tree=t)
+        self.assertRaises(ValueError, beta_diversity,
+                          'unweighted_unifrac', counts, otu_ids=otu_ids,
+                          tree=t)
+
+        # some otu_ids not present in tree
+        t = TreeNode.read(
+            StringIO(u'(((((OTU1:0.5,OTU2:0.5):0.5,OTU3:1.0):1.0):0.0,(OTU4:'
+                     u'0.75,OTU5:0.75):1.25):0.0)root;'))
+        counts = [1, 2, 3]
+        otu_ids = ['OTU1', 'OTU2', 'OTU42']
+        self.assertRaises(MissingNodeError, beta_diversity,
+                          'weighted_unifrac', counts, otu_ids=otu_ids, tree=t)
+        self.assertRaises(MissingNodeError, beta_diversity,
+                          'unweighted_unifrac', counts, otu_ids=otu_ids,
+                          tree=t)
+
+    def test_empty(self):
+        # array of empty vectors
+        actual = beta_diversity('euclidean',
+                                np.array([[], []], dtype=np.int64),
+                                ids=['a', 'b'])
+        expected_dm = DistanceMatrix([[0.0, 0.0], [0.0, 0.0]], ['a', 'b'])
+        npt.assert_array_equal(actual, expected_dm)
+
+        actual = beta_diversity('unweighted_unifrac',
+                                np.array([[], []], dtype=np.int64),
+                                ids=['a', 'b'], tree=self.tree1, otu_ids=[])
+        expected_dm = DistanceMatrix([[0.0, 0.0], [0.0, 0.0]], ['a', 'b'])
+        npt.assert_array_equal(actual, expected_dm)
+
+    def test_input_types(self):
+        actual_array = beta_diversity('euclidean',
+                                      np.array([[1, 5], [2, 3]]),
+                                      ids=['a', 'b'])
+        actual_list = beta_diversity('euclidean',
+                                     [[1, 5], [2, 3]], ids=['a', 'b'])
+        npt.assert_array_equal(actual_array, actual_list)
+
+    def test_euclidean(self):
+        actual_dm = beta_diversity('euclidean', self.table1, self.sids1)
+        self.assertEqual(actual_dm.shape, (3, 3))
+        npt.assert_almost_equal(actual_dm['A', 'A'], 0.0)
```

Comment:
We need a testing utility for asserting almost equal `DistanceMatrix` objects (or perhaps this can be implemented in a `.equals()` method). Can you add a TODO comment here for this? Not worth tackling now b/c `DistanceMatrix` is next on the list for redesign/refactoring after `skbio.alignment`.